modules/deploying-models-using-multiple-gpu-nodes.adoc

```diff
@@ -0,0 +1,253 @@
+
+.Procedure
+. In a terminal window, if you are not already logged in to your {openshift-platform} cluster as a cluster administrator, log in to the {openshift-platform} CLI as shown in the following example:
++
+[source]
+----
+$ oc login <openshift_cluster_url> -u <admin_username> -p <password>
+----
++
+
+. Select or create a namespace where you would like to deploy the model. In this example, `kserve-demo` is used and can be created using the following command:
++
+[source]
+----
+oc new-project kserve-demo
+----
++
+
+. Create the `vllm-multinode-runtime` ServingRuntime template as follows:
+
+.. Add the `vllm-multinode-runtime` ServingRuntime as a custom runtime:
++
+[source]
+----
+oc process vllm-multinode-runtime-template -n redhat-ods-applications|oc apply  -f -
+----
++
+
+.. From the namespace, define an InferenceService object in a YAML file called `vllm-multinode.yaml` with the following contents:
++
+[source]
+----
+... 
+ predictor:
+    model:
+      runtime: vllm-multinode-runtime
+      modelFormat:
+        name: vLLM
+      storageUri: pvc://llama-3-8b-pvc/hf/8b_instruction_tuned  
+    workerSpec: {} # Specifying workerSpec indicates that multi-node functionality will be used    
+----
++
+After you create the `vllm-multinode-runtime.yaml` file, the following configurations are added to the serving runtime:
++
+* `workerSpec.tensorParallelSize`: Determines how many GPUs are used per node. The GPU type count in both the head and worker node deployment resources is updated automatically. Ensure that the value of `workerSpec.tensorParallelSize` is at least `1`.
+* `workerSpec.pipelineParallelSize`: Determines how many nodes are involved in the deployment. This variable represents the total number of nodes, including both the head and worker nodes. Ensure that the value of `workerSpec.pipelineParallelSize` is at least `2`.
++
+
+. From the namespace where you would like to deploy the model, create a PVC for model storage and provide the name of your storage class. The storage class must be file storage.
++
+NOTE: If you have already configured a persistent volume claim (PVC), you can skip this step.
++
+[source]
+----
+kubectl apply -f - 
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: llama-3-8b-pvc
+spec:
+  accessModes:
+    - ReadWriteMany
+  volumeMode: Filesystem
+  resources:
+    requests:
+      storage: 50Gi
+  storageClassName: __<fileStorageClassName>__
+----
++
+
+. Download the model by exporting your token as an environment variable. In this example, `HF_TEST_TOKEN` is the exported token:
++
+[source]
+----
+export HF_TEST_TOKEN=__<token>__
+export MODEL=meta-llama/Meta-Llama-3-8B-Instruct
+
+curl -o download-model-to-pvc.yaml https://raw.githubusercontent.com/kserve/website/a5a8661de3c3bd55b1faa1acf9125d05b10bfefd/docs/modelserving/v1beta1/llm/huggingface/multi-node/download-model-to-pvc.yaml
+envsubst < download-model-to-pvc.yaml | kubectl create -f -
```

Comment:
should go after pvc (after project creation)

Download the model to the PVC, for example:
```
apiVersion: v1
kind: Pod
metadata:
  name: download-granite-8b-code
  labels:
    name: download-granite-8b-code
spec:
  volumes:
    - name: model-volume
      persistentVolumeClaim:
        claimName: granite-8b-code-claim
  restartPolicy: Never
  initContainers:
    - name: fix-volume-permissions
      image: quay.io/quay/busybox@sha256:92f3298bf80a1ba949140d77987f5de081f010337880cd771f7e7fc928f8c74d 
      command: ["sh"]
      args: ["-c", "mkdir -p /mnt/models/granite-8b-code-base && chmod -R 777 /mnt/models"]
      volumeMounts:
        - mountPath: "/mnt/models/"
          name: model-volume
  containers:
    - resources:
        requests:
          memory: 40Gi
      name: download-model
      imagePullPolicy: IfNotPresent
      image: quay.io/modh/kserve-storage-initializer@sha256:83ab292e87f2d30c5db4f756dac7d99d0a0252a4563f784f30a2d75cbbd46330
      args:
        - 's3://$(BUCKET_NAME)/granite-8b-code-base/'
        - /mnt/models/granite-8b-code-base
      env:
        - name: AWS_ACCESS_KEY_ID
          value: <id>
        - name: AWS_SECRET_ACCESS_KEY
          value: <secret>
        - name: BUCKET_NAME
          value: <bucket>
        - name: S3_USE_HTTPS
          value: "1"
        - name: AWS_ENDPOINT_URL
          value: <AWS endpoint>
        - name: awsAnonymousCredential
          value: 'false'
        - name: AWS_DEFAULT_REGION
          value: <region>
      volumeMounts:
        - mountPath: "/mnt/models/"
          name: model-volume
```